## Проверка данных

In [1]:
# импортируем библиотеки 
import pandas as pd 
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры 
db_config = {'user': 'praktikum_student', # имя пользователя  
             'pwd': 'Sdf4$2;d-d30pp', # пароль  
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',  
             'port': 6432, # порт подключения  
             'db': 'data-analyst-final-project-db'} 

# название базы данных 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']) 

# сохраняем коннектор 
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
def request(sql):
    return pd.io.sql.read_sql(sql, con = engine)

**Таблица books**

In [4]:
books = '''
          SELECT *
          FROM books
          LIMIT 5 
            '''
request(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица authors**

In [5]:
authors = '''
          SELECT *
          FROM authors
          LIMIT 5 
            '''
request(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица publishers**

In [6]:
publishers = '''
          SELECT *
          FROM publishers
          LIMIT 5 
            '''
request(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица ratings**

In [7]:
ratings = '''
          SELECT *
          FROM ratings
          LIMIT 5 
            '''
request(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица reviews**

In [8]:
reviews = '''
          SELECT *
          FROM reviews
          LIMIT 5 
            '''
request(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Исследование данных

### Количество книг, вышедших после 1 января 2000 года

In [9]:
book_count = '''
             SELECT COUNT(book_id)
             FROM books
             WHERE publication_date > '2000-01-01'
             '''
request(book_count)

count
0    819

После 1 января 2000 года вышло 819 книг.

### Количество обзоров и средняя оценка для каждой книги

In [10]:
count = '''
        SELECT b.book_id,
               title,
               COUNT(DISTINCT r.review_id) AS review,
               AVG(rt.rating) AS rating
        FROM books b
        LEFT JOIN reviews AS r ON b.book_id = r.book_id
        LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
        GROUP BY b.book_id
        ORDER BY COUNT(r.review_id) DESC
        '''
request(count)

book_id                                              title  review  \
0        948                            Twilight (Twilight  #1)       7   
1        750                The Hobbit  or There and Back Again       6   
2        673                             The Catcher in the Rye       6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...       6   
4        299  Harry Potter and the Chamber of Secrets (Harry...       6   
..       ...                                                ...     ...   
995      221                          Essential Tales and Poems       0   
996      387                               Leonardo's Notebooks       0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites       0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel       0   
999      808                            The Natural Way to Draw       0   

       rating  
0    3.662500  
1    4.125000  
2    3.825581  
3    4.414634  
4    4.287500  
..        ...  
995  4.000000  
996  4.000000  
997  5.000000  
998  3.666667  
999  3.000000  

[1000 rows x 4 columns]

Для каждой книги написано достаточно мало отзывов, максимальное количество отзывов - 7. Также есть книги без отзывов

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [11]:
publish = '''
          SELECT p.publisher,
                 COUNT(b.book_id) AS count             
          FROM books b
          LEFT JOIN publishers p ON b.publisher_id=p.publisher_id
          WHERE b.num_pages > 50
          GROUP BY p.publisher
          ORDER BY count DESC
          LIMIT 5
        '''
request(publish)

publisher  count
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25
3          Penguin Classics     24
4                    Bantam     19

Наибольшее чило книг толще 50 страниц выпустило издательство Penguin Books.

### Автор с самой высокой средней оценкой книг

In [12]:
author = '''
        SELECT a.author,
               AVG(r.rating) AS rating
        FROM books AS b
        INNER JOIN (SELECT book_id,
                   AVG(rating) AS rating
                   FROM ratings
                   GROUP BY book_id
                   HAVING COUNT(rating_id) > 50) AS r ON r.book_id=b.book_id
        LEFT JOIN authors AS a ON b.author_id=a.author_id
        GROUP BY a.author
        ORDER BY AVG(r.rating) DESC
            '''
request(author)

author    rating
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Автором с самым высоким рейтингом является Джоан Роулинг.

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
user = '''
        SELECT COUNT(rv.review_id) / COUNT(DISTINCT rv.username) AS avg_review
        FROM reviews AS rv
        INNER JOIN (SELECT username,
                    COUNT(rating_id)
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50) AS r ON r.username=rv.username
'''
request(user)

avg_review
0          24

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24. Пользователи оставляют отзывы не для всех оцененных книг.

## Вывод

- после 1 января 2000 года и на момент последних изменений в базе данных (который нам не известен) было выпущено 819 книг;
- в базе имеются книги без обзоров, а макимальное количество обзоров для книги - 7. В базе присутствуют книги с оценкой 5;
- наибольшее число книг толще 50 страниц выпустило издательство Penguin Books - 42 книги;
- автором с самым высоким рейтингом является Джоан Роулинг;
- среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24. Пользователи оставляют отзывы не для всех оцененных книг.